README.md
# CF Light API

[![Gem Version](https://badge.fury.io/rb/cf_light_api.svg)](http://badge.fury.io/rb/cf_light_api)

## What is this?

A super lightweight API for CloudFoundry. Why? Well the CF API contains all sorts of relations making it very heavy to query.

Having lots of API consumers for random scripts and dashboards makes it necessary to scale up the CF installation not to disrupt normal "cf cli operations".
So, lets just cache all the data we want in Redis for 5 minutes and serve that. \o/

This gem provides a single binary `cf_light_api`, which starts a small Sinatra app to serve the HTTP requests, and also starts a background worker via the Rufus Scheduler, which updates Redis with the data from CF every 5 minutes.

## API Endpoints

The API just reads a stringified JSON from Redis and serves it under the following endpoints.

### Apps

There are currently two endpoints, the original (and now deprecated) "v1" and a new "v2" which offers more details and better error handling.

#### V2 - Current

`GET /v2/apps`

##### (Optional) Filtering by Org name

`GET /v2/apps/<org_name>`

##### Response

An array of JSON documents for all applications in the configured CF environment. If you provide an `org` parameter, the list will be filtered to contain only applications belonging to the given `org`. The majority of the document is identical to the response you would receive from a standard `/v2/apps` request made against the main CF API, but with some additional fields added on top:

* `created_at` - When the app was originally created, taken from the metadata.

* `updated_at` - When the app was last modified, taken from the metadata.

* `guid` - The app GUID, taken from the metadata.

* `instances` - An array of all instances associated with this application, their state and usage statistics

* `routes` - An array of all routes assigned to your app instances (also exposed via the `uris` attribute of each member of the `instances` attribute).

* `meta` - A document containing any error messages capturing during processing of this application.

* `stack` - The name of the container stack being used by the application, as a string.

* `space` - The name of the space this app belongs to, as a string.

* `org` - The name of the org this app belongs to, as a string.

* `environment_json` - An array of environment variables exposed to this application, if enabled as documented [here](#gathering-environment-variables).

* `buildpack_name` - A field containing the name of the buildpack in use. Its value is equal either to the `buildpack` attribute of the associated app or should this be `none` (autodetected buildpack) to the buildpack name as resolved from the `v2/buildpacks/<guid>` endpoint using the `detected_buildpack_guid` field as `guid`.

Each document has the following structure:

```json
{
  "name": "app_name",
  "production": false,
  "space_guid": "c0af44b8-8b51-4db5-927e-ccad2e6dab54",
  "stack_guid": "7c5b664c-e9b8-457b-83a4-6092b7372494",
  "buildpack": null,
  "detected_buildpack": "Ruby",
  "detected_buildpack_guid": "44ec3a97-0d94-4ebb-ad33-e9ee837515bd",
  "environment_json": {},
  "memory": 64,
  "instances": [
    {
      "state": "RUNNING",
      "stats": {
        "name": "app_name",
        "uris": [
          "app_name.yourdomain.com"
        ],
        "host": "ip address",
        "port": "port number",
        "uptime": 2889203,
        "mem_quota": 67108864,
        "disk_quota": 1073741824,
        "fds_quota": 16384,
        "usage": {
          "time": "2018-01-19T15:54:48.254719257Z",
          "cpu": 0.00012013292649106141,
          "mem": 43802624,
          "disk": 78118912
        }
      }
    }
  ],
  "disk_quota": 1024,
  "state": "STARTED",
  "version": "45c81548-3d3d-4152-bbc5-6b86eba4d5df",
  "command": null,
  "console": false,
  "debug": null,
  "staging_task_id": "ba623c5c-18e1-4d6e-b331-aedf244cb493",
  "package_state": "STAGED",
  "health_check_type": "port",
  "health_check_timeout": null,
  "health_check_http_endpoint": null,
  "staging_failed_reason": null,
  "staging_failed_description": null,
  "diego": true,
  "docker_image": null,
  "docker_credentials": {
    "username": null,
    "password": null
  },
  "package_updated_at": "2017-06-13T12:19:37Z",
  "detected_start_command": "bundle exec rackup config.ru -p $PORT",
  "enable_ssh": false,
  "ports": [
    8080
  ],
  "space_url": "/v2/spaces/c0af44b8-8b51-4db5-927e-ccad2e6dab54",
  "stack_url": "/v2/stacks/7c5b664c-e9b8-457b-83a4-6092b7372494",
  "stack": "cflinuxfs2",
  "routes_url": "/v2/apps/ba623c5c-18e1-4d6e-b331-aedf244cb493/routes",
  "routes": [
    "app_name.yourdomain.com"
  ],
  "events_url": "/v2/apps/ba623c5c-18e1-4d6e-b331-aedf244cb493/events",
  "service_bindings_url": "/v2/apps/ba623c5c-18e1-4d6e-b331-aedf244cb493/service_bindings",
  "route_mappings_url": "/v2/apps/ba623c5c-18e1-4d6e-b331-aedf244cb493/route_mappings",
  "created_at": "2015-04-22T12:07:56Z",
  "updated_at": "2016-12-20T10:02:47Z",
  "guid": "ba623c5c-18e1-4d6e-b331-aedf244cb493",
  "running": true,
  "environment_variables": {},
  "meta": {
    "error": false
  },
  "buildpack_name": "buildpack_name",
  "space": "space name",
  "org": "org name"
}
```

##### Errors

If there are any errors processing a given application, details on the error will be captured and exposed in the `meta` attribute, for example:

```
"meta": {
  "error": true,
  "type": "CFResponseError",
  "message": "Code 200003: CF-AppStoppedStatsError - Could not fetch stats for stopped app: <name of app here>",
  "backtrace": [
    "etc...", "etc...", "etc..."
  ]
}
```

If there are no errors, the `meta` attribute will instead look like this:

```
"meta": {
  "error": false
}
```

##### Notes

* The `running` attribute is a boolean and will be `true` if there is at least one instance of your app with the state of `RUNNING`, or `false` otherwise.

* Memory, disk quota and usage figures are given in bytes.

* If the buildpack is not known, the `buildpack` attribute will be `null`. However, the `buildpack_name` field will be populated using the buildpack `detected_buildpack_guid` value.

* If the last uploaded time is not known, the `last_uploaded` attribute will be `null`.

* The `diego` attribute is a boolean which will be `true` if the application is running on a Diego Cell, or `false` if running on a DEA Node. See the Cloud Foundry [docs](https://docs.cloudfoundry.org/concepts/diego/dea-vs-diego.html) for more information on these two architectures.

* There is an optional `environment_json` attribute which only appears if the feature is enabled - please see the [Gathering Environment Variables](#gathering-environment-variables) section below for more information.

#### V1 - Deprecated

`GET /v1/apps`

##### (Optional) Filtering by Org name

`GET /v1/apps/<org_name>`

##### Response

An array of JSON documents for all applications in the configured CF environment. If you provide an `org` parameter, the list will be filtered to contain only applications belonging to the given `org`. Each document has the following structure:

```json
{
  "guid": "app GUID",
  "name": "app_name",
  "org": "org name",
  "space": "space name",
  "stack": "lucid64",
  "buildpack": "https://github.com/cloudfoundry/ruby-buildpack.git",
  "diego": false,
  "docker": false,
  "docker_image": null,
  "routes": [
    "app_name.yourdomain.com"
  ],
  "data_from": "unix timestamp of last update",
  "last_uploaded": "2016-02-04 15:21:25 +0000",
  "state": "STARTED",
  "running": true,
  "instances": [
    {
      "state": "RUNNING",
      "stats": {
        "name": "app_name",
        "uris": [
          "app_name.yourdomain.com"
        ],
        "host": "ip address",
        "port": "port number",
        "uptime": 1979434,
        "mem_quota": 268435456,
        "disk_quota": 268435456,
        "fds_quota": 16384,
        "usage": {
          "time": "2015-02-27 16:52:35 +0000",
          "cpu": 0.0,
          "mem": 134217728,
          "disk": 134217728
        }
      }
    }
  ],
  "error": null
}
```

##### Notes

* The `running` attribute may contain `true`, `false` or `error`. Applications in the latter state will have further information about the problem in the `error` attribute, which is `null` at all other times.

* Memory, disk quota and usage figures are given in bytes.

* If the buildpack is not known, the `buildpack` attribute will be `null`.

* If the last uploaded time is not known, the `last_uploaded` attribute will be `null`.

* The `diego` attribute is a boolean which will be `true` if the application is running on a Diego Cell, or `false` if running on a DEA Node. See the Cloud Foundry [docs](https://docs.cloudfoundry.org/concepts/diego/dea-vs-diego.html) for more information on these two architectures.

* The `docker` attribute is a boolean which will be `true` if the application was deployed from a Docker image, or `false` if it was deployed using a buildpack. If `true`, the `docker_image` attribute will then show the Docker image used, otherwise it will be `null`.

* There is an optional `environment_variables` attribute which only appears if the feature is enabled - please see the section of Gathering Environment Variables below for more information.

### Organisations

`GET /v1/orgs`

#### (Optional) Filtering by Org GUID

`GET /v1/orgs/<guid>`

#### Response

An array of JSON documents, for all organisations in your CF environment. If you provide an `guid` parameter, the list will be filtered to contain only the organisation with the given `guid`. Each document has the following structure:

```json
{
  "guid": "org GUID",
  "name": "my_org_name",
  "quota": {
    "total_services": 50,
    "memory_limit": 10737418240
  }
}
```

##### Notes

* Memory limits for each org are given in bytes.

### Last Updated

`GET /v1/last_updated`

####

A single JSON document showing the last time the data was updated by the worker:

```json
{
  "last_updated": "2015-11-07 01:25:28 +0000"
}
```

If the data was not updated within a configurable time period, the HTTP status will be `503 Service Unavailable`.

### Info

`GET /v1/info`

#### Response

A single JSON document showing various pieces of information about the CF installation. This data comes unmodified from CF's own `/v2/info` endpoint.

```json
{
  "name": "vcap",
  "build": "2222",
  "support": "http://support.cloudfoundry.com",
  "version": 2,
  "description": "Cloud Foundry sponsored by Pivotal",
  "authorization_endpoint": "http://localhost:8080/uaa",
  "token_endpoint": "http://localhost:8080/uaa",
  "min_cli_version": "6.20.0",
  "min_recommended_cli_version": "6.21.0",
  "api_version": "2.94.0",
  "app_ssh_endpoint": "ssh.system.domain.example.com:2222",
  "app_ssh_host_key_fingerprint": "47:0d:d1:c8:c3:3d:0a:36:d1:49:2f:f2:90:27:31:d0",
  "app_ssh_oauth_client": null,
  "routing_endpoint": "http://localhost:3000",
  "doppler_logging_endpoint": "wss://doppler.vcap.me:4443",
  "logging_endpoint": "ws://loggregator.vcap.me:80"
}
```

## Worker

The worker basically gets all the data we want from the real API every 5 mins, puts in Redis and sleeps. The worker runs in a background thread via the Rufus Scheduler and is automatically started as part of the API. There is basic locking implemented via Redis to allow you to run multiple instances of `cf_light_api` alongside one another, without duplicating work.

## Usage

1. You must first set the following environment variables:
```bash

CHANGELOG.md

# 3.1.2 (March 29, 2018)

Bugfixes:

- (3.1.2.pre - March 22, 2018) Handle case when both `detected_buildpack_guid` and `buildpack` are `nil` due to a bad application state (ie.) bits not uploaded.

# 3.1.1

Bugfixes:

-  (3.1.1 - March 21, 2018) - `3.1.0` failed to include the relevant commits and has been replaced by 3.1.1 (functionally identical)

# 3.1.0 (Yanked release - replaced by `3.1.1`)

Refactoring:

  - (3.1.0.pre2 - March 21, 2018): Refactor `buildpack_name` to conditional assignment to avoid redundancy.

Features:

  - (3.1.0.pre1 - March 21, 2018): Added new `buildpack_name` field to the `v2` manifest in the `/v2/apps` endpoint. This makes sure the buildpack name is populated when autodetected (in this case the `buildpack` field defaults to `None`).

# 3.0.0

We've refactored the worker to improve error handling, and to make it possible to capture errors from CF during processing of each app, so we can propagate this out to the consumers.

Furthermore, we now also capture the entire response from Cloud Foundry, for each app, rather than mapping only a subset of the attributes, this should allow consumers to make more informed decisions with the data.

Contains several features and bugfixes, tested via several pre-release versions:

Features (3.0.0.pre1 - January 22, 2018):

  - New `/v2/apps` endpoint which returns all the app data available from CF, rather than a subset of the fields as before.

  - New `meta` attribute present for each app in `/v2/apps`, which contains details of any errors encountered when processing the relevant app.

  - Adjusted default log levels to be less noisy, and added an option to enable debug logging via a `DEBUG` environment variable when needed.

Deprecated:

  - The `/v1/apps` endpoint remains, but should now be considered deprecated.

Bugfixes:

  - (3.0.0.pre8 - January 29, 2018): Ensure the `routes` and `instances` attributes are once again always present in the `/v1/apps` endpoint.

  - (3.0.0.pre7 (January 24, 2018))  - In order to avoid an exception, we only pass through running instances when formatting usage stats to be sent to Graphite.

  - 3.0.0.pre6 (January 23, 2018) - Don't attempt to send formatted instance stats to Graphite when there are no instances.

  - 3.0.0.pre5 (January 23, 2018) - Fixed a scoping issue when checking if there is at least one running instance.

  - 3.0.0.pre4 (January 23, 2018) - Restoring Graphite functionality, which was disabled in the previous pre-release versions pre1, pre2 and pre3.

  - 3.0.0.pre3 (January 23, 2018) - We no longer try to gather app instance stats, if the app state is 'STOPPED'.

  - 3.0.0.pre2 (January 23, 2018) - Fixes a regression to do with Graphite metrics when certain inline-relations are not present in the app object, for example "stack".

# 2.6.0 (December 4th, 2017)

Contains several features and bugfixes, tested via the following pre-release versions:

## 2.6.0.pre4

Features:

  - Upgraded some of our dependencies to the latest versions.

## 2.6.0.pre3

Bugfixes:

  - Ensure the response from CF is a valid JSON object, and correctly handle this as an error if it is not.
  - Added error handling to the domain lookup logic for app routes.
  - Stacktraces are now logged along with the error message, when rescuing from StandardError.

## 2.6.0.pre2 (November 29, 2017)

Bugfixes:

  - Fixed a bug when sending worker update duration to Graphite.

## 2.6.0.pre1 (November 29, 2017)

Features:

  - Reintroduced the parallelisation option when gathering app data, to speed things up for larger Cloud Foundry deployments. Defaults to the current behaviour of a single thread.
  - The duration of the worker update is now sent to Graphite under a new key `cf_light_api.${cf_environment}.update_duration`. It is given in seconds, or `0` if the timeout was reached during the update.

Bugfixes:

  - Fixed a bug where TimeoutErrors were being caught at the wrong level, and causing an `error` state to be set for apps which were actually running, resulting in the API returning incorrect data.

# 2.5.1 (November 21, 2017)

Features:

  - Adds an additional endpoint `/v1/info` which returns the unmodified response from CF's own `/v2/info` endpoint, containing various useful bits of information about your CF installation.

Bugfixes:

  - Sometimes route gathering would fail for apps in certain edge case states, this should now be handled gracefully instead of causing the data update to fail.

# 2.5.0 (June 19, 2017)

Features:

  - Adds an additional opt-in attribute to the response for `/v1/apps`:
    * `environment_variables`
      A dictionary containing an app's environment variables. This feature must be enabled explicitly and can also be configured with a whitelist to protect sensitive environment variables from being exposed.

# 2.4.0 (January 16, 2017)

Features:

  - Send organisation quota details to Graphite (when enabled).

# 2.3.0 (January 11, 2017)

Features:

  - Changed the Graphite key path to include the CF environment name, so users with more than one Cloud Foundry deployment can group metrics by deployment.

# 2.2.1.pre1 (November 30, 2016)

Bugfixes:

  - Fix an issue with sanitising the app name for Graphite if it included `.`, which was also altering the app name when reported by the API itself.

# 2.2.0.pre1 (November 24, 2016)

Features:

  - Instrumentation regarding the API and the backend worker is now available thanks to new integration with New Relic.

# 2.1.0 (August 9, 2016)

Features:


lib/cf_light_api/worker.rb
require 'cfoundry'
require 'json'
require 'rufus-scheduler'
require 'redlock'
require 'logger'
require 'graphite-api'
require 'date'
require 'parallel'

class CFLightAPIWorker
  if ENV['NEW_RELIC_LICENSE_KEY']
    require 'newrelic_rpm'
    include NewRelic::Agent::Instrumentation::ControllerInstrumentation
    include NewRelic::Agent::MethodTracer
  end

  ENVIRONMENT_VARIABLES_WHITELIST = (ENV['ENVIRONMENT_VARIABLES_WHITELIST'] || '').split(',').collect(&:strip)

  def initialize
    @logger = Logger.new(STDOUT)

    if ENV['DEBUG']
      @logger.level = Logger::DEBUG
    else
      @logger.level = Logger::INFO
    end

    @logger.formatter = proc do |severity, datetime, progname, msg|
       "#{datetime} [cf_light_api:worker]: #{msg}\n"
    end

    ['CF_API', 'CF_USER', 'CF_PASSWORD'].each do |env|
      unless ENV[env]
        @logger.info "Error: please set the '#{env}' environment variable."
        exit 1
      end
    end

    # If either of the Graphite settings are set, verify that they are both set, or exit with an error. CF_ENV_NAME is used
    # to prefix the Graphite key, to allow filtering by environment if you run more than one.
    if ENV['GRAPHITE_HOST'] or ENV['GRAPHITE_PORT']
      ['GRAPHITE_HOST', 'GRAPHITE_PORT', 'CF_ENV_NAME'].each do |env|
        unless ENV[env]
          @logger.info "Error: please also set the '#{env}' environment variable to enable exporting to Graphite."
          exit 1
        end
      end
    end

    update_interval = (ENV['UPDATE_INTERVAL'] || '5m').to_s # If you change the default '5m' here, also remember to change the default age validity in sinatra/cf_light_api.rb:31
    update_timeout  = (ENV['UPDATE_TIMEOUT']  || '5m').to_s

    @update_threads  = (ENV['UPDATE_THREADS'] || 1).to_i

    @lock_manager = Redlock::Client.new([ENV['REDIS_URI']])
    @scheduler    = Rufus::Scheduler.new

    @logger.info "Update interval: '#{update_interval}'"
    @logger.info "Update timeout:  '#{update_timeout}'"
    @logger.info "Update threads:  '#{@update_threads}'"

    if ENV['GRAPHITE_HOST'] and ENV['GRAPHITE_PORT']
      @logger.info "Graphite server: #{ENV['GRAPHITE_HOST']}:#{ENV['GRAPHITE_PORT']}"
    else
      @logger.info 'Graphite server: Disabled'
    end

    @scheduler.every update_interval, :first_in => '5s', :overlap => false, :timeout => update_timeout do
      update_cf_data
    end
  end

  def cf_rest(path, method='GET')
    @logger.debug "Making #{method} request for #{path}..."

    resources = []
    options   = {:accept => :json}
    response  = @cf_client.base.rest_client.request(method, path, options)[1][:body]

    begin
      response = JSON.parse(response)
      if response['error_code']
        raise CFResponseError.new("Code #{response['code']}, #{response['error_code']} - #{response['description']}")
      end

    rescue Rufus::Scheduler::TimeoutError => e
      raise e
    rescue JSON::ParserError => e
      @logger.error "Error parsing JSON response from #{method} #{path}: #{e.message}"
      @logger.trace e.backtrace
      @logger.trace response
      raise e
    rescue CFoundry => e
      @logger.error "CFoundry error making #{method} #{path}: #{e.message}"
      @logger.trace e.backtrace
      @logger.trace response
      raise e
    rescue CFResponseError => e
      @logger.error "CF API returned a response with an error document for #{method} #{path}: #{e.message}"
      @logger.trace e.backtrace
      @logger.trace response
      raise e
    rescue StandardError => e
      @logger.error "General error making #{method} #{path}: #{e.message}"
      @logger.trace e.backtrace
      @logger.trace response
      raise e
    end

    # Some endpoints return a 'resources' array, others are flat, depending on the path.
    if response['resources']
      resources << response['resources']
    else
      resources << response
    end

    # Handle the pagination by recursing over myself until we get a response which doesn't contain a 'next_url'
    # at which point all the resources are returned up the stack and flattened.
    resources << cf_rest(response['next_url'], method) unless response['next_url'] == nil
    resources.flatten
  end

  def get_client(cf_api=ENV['CF_API'], cf_user=ENV['CF_USER'], cf_password=ENV['CF_PASSWORD'])
    client = CFoundry::Client.get(cf_api)
    client.login({:username => cf_user, :password => cf_password})
    client
  end

  def send_instance_usage_data_to_graphite(instance_stats, org, space, app_name)
    sanitised_app_name = app_name.gsub ".", "_" # Some apps have dots in the app name which breaks the Graphite key path

    instance_stats.each_with_index do |instance_data, index|
      graphite_base_key = "cf_apps.#{ENV['CF_ENV_NAME']}.#{org}.#{space}.#{sanitised_app_name}.#{index}"
      @logger.debug "  Exporting app instance \##{index} usage statistics to Graphite, path '#{graphite_base_key}'"

      # Quota data
      ['mem_quota', 'disk_quota'].each do |key|
        @logger.trace "#{graphite_base_key}.#{key} => #{instance_data['stats'][key]}"
        @graphite.metrics "#{graphite_base_key}.#{key}" => instance_data['stats'][key]
      end

      # Usage data
      ['mem', 'disk', 'cpu'].each do |key|
        @logger.trace "#{graphite_base_key}.#{key} => #{instance_data['stats']['usage'][key]}"
        @graphite.metrics "#{graphite_base_key}.#{key}" => instance_data['stats']['usage'][key]
      end
    end
  end

  def send_org_quota_data_to_graphite(org_name, quota)
    graphite_base_key = "cf_orgs.#{ENV['CF_ENV_NAME']}.#{org_name}"
    @logger.debug "  Exporting org quota statistics to Graphite, path '#{graphite_base_key}'"

    quota.keys.each do |key|
      @logger.trace "#{graphite_base_key}.quota.#{key} => #{quota[key]}"
      @graphite.metrics "#{graphite_base_key}.quota.#{key}" => quota[key]
    end
  end

  def send_cf_light_api_update_time_to_graphite seconds
    graphite_key = "cf_light_api.#{ENV['CF_ENV_NAME']}.update_duration"
    @logger.info "Exporting CF Light API update time to Graphite, path #{graphite_key} => #{seconds.round}"
    @graphite.metrics "#{graphite_key}" => seconds.round
  end

  def put_in_redis(key, data)
    REDIS.set key, data.to_json
  end

  def format_duration(elapsed_seconds)
    seconds = elapsed_seconds % 60
    minutes = (elapsed_seconds / 60) % 60
    hours   = elapsed_seconds / (60 * 60)
    format("%02d hrs, %02d mins, %02d secs", hours, minutes, seconds)
  end

  def update_domains
    @domains = cf_rest('/v2/domains?results-per-page=100')
  end

  def get_buildpacks_by_guid
    buildpacks = cf_rest('/v2/buildpacks?results-per-page=100')
    buildpacks_by_guid = buildpacks.map { |buildpack| [buildpack['metadata']['guid'], buildpack] }.to_h
  end

  def find_domain_for_route route
    return @domains.find{|a_domain| a_domain['metadata']['guid'] == route['entity']['domain_guid']}
  end

  def format_routes_for_app app
    # The app object passed in here should contain a "routes" attribute, fetched as part of the original request to CF (with inline-relation gathering enabled)
    # and it will look something like this:
    # "routes"=>
    # [{"metadata"=>{"guid"=>"afea5690-fb93-451a-9610-2d524d36e35f", "url"=>"/v2/routes/afea5690-fb93-451a-9610-2d524d36e35f", "created_at"=>"2015-03-11T12:20:22Z", "updated_at"=>"2015-03-11T12:20:22Z"},
    #   "entity"=>
    #    {"host"=>"hostname_here",
    #     "path"=>"",
    #     "domain_guid"=>"f13e6864-537e-41bb-b46c-f3810dbf7c84",
    #     "space_guid"=>"c0af44b8-8b51-4db5-927e-ccad2e6dab54",
    #     "service_instance_guid"=>nil,
    #     "port"=>nil,
    #     "domain_url"=>"/v2/shared_domains/f13e6864-537e-41bb-b46c-f3810dbf7c84",
    #     "space_url"=>"/v2/spaces/c0af44b8-8b51-4db5-927e-ccad2e6dab54",
    #     "apps_url"=>"/v2/routes/afea5690-fb93-451a-9610-2d524d36e35f/apps",
    #     "route_mappings_url"=>"/v2/routes/afea5690-fb93-451a-9610-2d524d36e35f/route_mappings"}},
    # If we don't receive that child attribute, (perhaps the app was being staged or didn't have any routes yet) we make another request to CF to try
    # and fetch them before giving up and just returning an empty array.

    routes = []
    if app['entity']['routes'] == nil
      # We have no routes data inlined with the app entity, so let's try to retrieve them directly from CF
      routes = cf_rest(app['entity']['routes_url'])
    else
      # Routes were already retrieved as an inline-relation, so just use those...
      routes = app['entity']['routes']
    end

    routes.collect do |route|
      host   = route['entity']['host']
      path   = route['entity']['path']

      domain = find_domain_for_route(route)
      if domain == nil
        # The domain doesn't exist, this could be due to a race condition, so let's update the list and try again
        update_domains()
        domain = find_domain_for_route(route)
        if domain == nil
          # If we can't determine the domain associated with this route, raise an error as we can't guarantee the state is correct here,
          # it shouldn't be possible to get a route back from CF with a domain GUID that doesn't exist, as that route would be invalid.
          raise "Unable to find domain #{route['entity']['domain_guid']} for route #{route['metadata']['guid']}."
        end
      end
      "#{host}.#{domain['entity']['name']}#{path}"
    end
  end

  def filtered_environment_variables env_vars
    if ENVIRONMENT_VARIABLES_WHITELIST.any?
      return ENVIRONMENT_VARIABLES_WHITELIST.inject({}) do |filtered, key|
        filtered[key] = env_vars[key] if env_vars[key]
        filtered
      end

    else
      return env_vars
    end
  end

  def format_orgs orgs
    return orgs.map do |org|
      quota = @quotas.find{|a_quota| a_quota['metadata']['guid'] == org['entity']['quota_definition_guid']}

      quota = {
        :total_services => quota['entity']['total_services'],
        :total_routes   => quota['entity']['total_routes'],
        :memory_limit   => quota['entity']['memory_limit'] * 1024 * 1024
      }

      send_org_quota_data_to_graphite(org['entity']['name'], quota) if @graphite

      {
        :guid => org['metadata']['guid'],
        :name  => org['entity']['name'],
        :quota => quota
      }
    end
  end

  def get_v1_base_data app
    # Format the app data in the expected format for the /v1/apps endpoint, to remain compatible.

    # Find the org for this app, using the org GUID from the space. Relationship: Apps belong to spaces, and spaces belong to orgs.
    space = @spaces.find{|a_space| a_space['metadata']['guid'] == app['entity']['space_guid']}
    org = @orgs.find{|an_org| an_org['metadata']['guid'] == space['entity']['organization_guid']}

    {
      :buildpack     => app['entity']['buildpack'],
      :data_from     => Time.now.to_i,
      :diego         => app['entity']['diego'],
      :docker        => app['entity']['docker_image'] ? true : false,
      :docker_image  => app['entity']['docker_image'],
      :guid          => app['metadata']['guid'],
      :last_uploaded => app['metadata']['updated_at'] ? DateTime.parse(app['metadata']['updated_at']).strftime('%Y-%m-%d %T %z') : nil,
      :name          => app['entity']['name'],
      :org           => org['entity']['name'],
      :space         => space['entity']['name'],
      :stack         => app['entity']['stack']['entity']['name'],
      :state         => app['entity']['state'],
      :instances     => [],
      :routes        => []
    }
  end

  def update_cf_data
    @cf_client = nil
    @graphite  = GraphiteAPI.new(graphite: "#{ENV['GRAPHITE_HOST']}:#{ENV['GRAPHITE_PORT']}") if ENV['GRAPHITE_HOST'] and ENV['GRAPHITE_PORT'] and ENV['CF_ENV_NAME']

    begin
      @lock_manager.lock("#{ENV['REDIS_KEY_PREFIX']}:lock", 5*60*1000) do |lock|
        if lock
          start_time = Time.now
          @logger.info "Updating data..."
          @cf_client = get_client() # Ensure we have a fresh auth token...

          @cf_info  = cf_rest('/v2/info').first

          @apps = cf_rest('/v2/apps?results-per-page=100&inline-relations-depth=1&include-relations=routes,stack')
          @spaces  = cf_rest('/v2/spaces?results-per-page=100')
          @buildpacks = get_buildpacks_by_guid() # Sets @buildpacks to a map of buildpack resources indexed by guid

          update_domains() # Sets @domain by hitting the CF API

          # Orgs
          @orgs          = cf_rest('/v2/organizations?results-per-page=100')
          @quotas        = cf_rest('/v2/quota_definitions?results-per-page=100')
          formatted_orgs = format_orgs @orgs

          v1_data = []
          v2_data = []

          Parallel.each(@apps, :in_threads => @update_threads) do |app|
            begin
              # Formats the base data compatible with the v1 endpoint
              v1_document = get_v1_base_data(app)

              # New format base data for the v2 endpoint
              v2_document                          = app['entity'].dup
              v2_document['environment_json']      = {} # The environment JSON will have been duplicated from the app entity, so we need to blank it here, as it will be re-populated later if EXPOSE_ENVIRONMENT_VARIABLES is true.
              v2_document['created_at']            = app['metadata']['created_at']
              v2_document['updated_at']            = app['metadata']['updated_at']
              v2_document['guid']                  = app['metadata']['guid']
              v2_document['instances']             = []
              v2_document['routes']                = []
              v2_document['meta']                  = { 'error' => false }

              # Add buildpack_name as a top level string attribute and looks it up using its guid when the buildpack field is null
              buildpack_name = app['entity']['buildpack']
              buildpack_guid = app['entity']['detected_buildpack_guid']

              v2_document['buildpack_name'] =
                if @buildpacks.has_key?(buildpack_guid) and (buildpack_name.nil? or buildpack_name.empty?)
                    @buildpacks[buildpack_guid]['entity']['name']
                else
                  buildpack_name
                end

              # Add space, stack and org names as a top level string attribute for ease of use:
              v2_document['stack'] = app['entity']['stack']['entity']['name']

              # Get the org name from the app's space - relationship: an app belongs to a space, and a space belongs to an org.
              space = @spaces.find{|a_space| a_space['metadata']['guid'] == app['entity']['space_guid']}
              org = @orgs.find{|an_org| an_org['metadata']['guid'] == space['entity']['organization_guid']}

              v2_document['space'] = space['entity']['name']
              v1_document['org'] = org['entity']['name']
              v2_document['org'] = org['entity']['name']

              # Gather and filter environment variable JSON if the feature is enabled:
              if ENV['EXPOSE_ENVIRONMENT_VARIABLES'] == 'true' then
                env_vars = filtered_environment_variables( app['entity']['environment_json'] )
                v1_document['environment_variables'] = env_vars
                v2_document['environment_json']      = env_vars
              end

              routes = format_routes_for_app(app)
              v1_document['routes'] = routes
              v2_document['routes'] = routes

              # Try to gather app instance stats, unless the app is stopped...
              unless app['entity']['state'] == 'STOPPED'
                response = cf_rest("/v2/apps/#{app['metadata']['guid']}/stats")
                instances = response.first.map{|key,value|value}
                v1_document['instances'] = instances
                v2_document['instances'] = instances
              end

              # We consider an app to be "running" if there is at least one app instance available with a state of "RUNNING"
              running = false
              running_instances = []
              if v2_document['instances'].any?
                running_instances = v2_document['instances'].select{|instance| instance['state'] == 'RUNNING'}
                running = true if running_instances.any?
              end
              v1_document['running'] = running
              v2_document['running'] = running

              if @graphite
                if running_instances.any?
                  send_instance_usage_data_to_graphite(running_instances, v2_document['org'], v2_document['space'], v2_document['name'])
                end
              end

            rescue Rufus::Scheduler::TimeoutError => e
              raise e
            rescue CFoundry, CFResponseError, StandardError => e
              v1_document['running']           = "error"
              v1_document['error']             = "#{e.message}"

              v2_document['meta']              = {}
              v2_document['meta']['error']     = true
              v2_document['meta']['type']      = e.class
              v2_document['meta']['message']   = e.message
              v2_document['meta']['backtrace'] = e.backtrace
            end

            v1_data << v1_document
            v2_data << v2_document
          end

          # Sanity check - do we have the expected quantity of data? This shouldn't happen as the `parallel` gem should handle
          # sharing and modifying variables for us when using threads.
          if @apps.count != v1_data.count or @apps.count != v2_data.count
            raise "V1 and V2 app counts don't match after processing!"
          end

          put_in_redis "#{ENV['REDIS_KEY_PREFIX']}:info", @cf_info
          put_in_redis "#{ENV['REDIS_KEY_PREFIX']}:orgs", formatted_orgs
          put_in_redis "#{ENV['REDIS_KEY_PREFIX']}:apps", v1_data
          put_in_redis "#{ENV['REDIS_KEY_PREFIX']}:apps:v2", v2_data

          put_in_redis "#{ENV['REDIS_KEY_PREFIX']}:last_updated", {:last_updated => Time.now}

          elapsed_seconds = Time.now.to_f - start_time.to_f

          send_cf_light_api_update_time_to_graphite(elapsed_seconds) if @graphite
          @logger.info "Update completed in #{format_duration(elapsed_seconds)}..."

          @lock_manager.unlock(lock)
          @cf_client.logout
        else
          @logger.info "Update already running in another instance!"
        end
      end
    rescue Rufus::Scheduler::TimeoutError
      Parallel::Kill
      @cf_client.logout
      @logger.info 'Data update took too long and was aborted, waiting for the lock to expire before trying again...'
      send_cf_light_api_update_time_to_graphite(0) if @graphite
    rescue StandardError => e
      @logger.info "Unable to complete update due to #{e.class}: #{e.message}"
      @logger.error e.backtrace
    end
  end

  if ENV['NEW_RELIC_LICENSE_KEY']
    add_transaction_tracer :update_cf_data, category: :task
    add_method_tracer :update_cf_data
  end

end

CFLightAPIWorker.new
